### ISSUE TITLE
Two bugs when unsubscribing from subscription

### ISSUE THREAD

**Issue by mpoeter** [absinthe-graphql/absinthe-socket#15]
The socket-apollo-link automatically multiplexes multiple subscriptions, i.e., when I have several identical subscription-queries only one such query is sent over the socket. This query has an associated notifier, which in turn holds a a list of observers that will be notified when a corresponding subscription is received.

This is nice, but unfortunately there are two issues with the current implementation.

First, suppose there are two identical subscriptions resulting in a single notifier with two observers. When unsubscribe is called for one of the observers is called, the whole notifier is removed, effectively disabling notifications for all observers in this notifier.
To fix this, a notifier must only be removed once all observers are removed.

Second, there is a race condition when unsubscribing from a subscription. The unsubscribe function for the observer (in `notifierToObservable`in `@absinthe/socket-apollo-link/src/createAbsintheSocketLink.js`) calls `cancel`to remove the notifier. This triggers an unsubscribe message to be send over the socket, and when the reply is received the notifier gets removed.
However, it can happen that a new subscription adds another observer to the notifier *before* the reply is received. Even if the notifier would not be removed due to the previous fix (since the observer list is not empty), the subscription would be invalid since it has been removed on the server side.
To fix this, the notifier must be removed *before* the unsubscribe message is sent.

**Comment by mgtitimoli** [comment]
Hi @mpoeter,

Thanks for having taken the time to report this.

The first issue you described is already handled, based on what you wrote, it seems that you might be calling cancel, instead of unobserve, could it be?

Regarding with the second one, I'm already aware of it, and I'm currently working on a release which will include code to handle this.

**Comment by mpoeter** [comment]
Thanks for the quick reply @mgtitimoli 

Actually the first issue still remains. We use the socket-apollo-link package which provides an Observable to observe incoming subscriptions. When the observable is no longer needed we call unsubscribe on it, but the unsubscribe function defined in `notifierToObservable` in `@absinthe/socket-apollo-link/src/createAbsintheSocketLink.js` itself calls `cancel`, thus removing the notifier and effectively disabling notifications for all observers.

**Comment by mgtitimoli** [comment]
oh, I see, I was reading the socket code directly without taking much into account what was going on in the apollo package, that within it is where the error is located.

In any case this border case will also be handled in the next release, as is exactly what I'm working on now, that's basically improve how cancellations are handled, great you made me notice about this, as I was only paying attention to the other error you mention.

So, long story short, I will try to add a patch for the first error this week, but the proper fix, this is, the next release, will be ready by the end of the month.

Thanks again @mpoeter.

**Comment by mpoeter** [comment]
Sounds great! Thanks again for the prompt reply and the quick fix - much appreciated!

**Comment by mpoeter** [comment]
*push*

Is anybody still maintaining this repository?

**Comment by rjacobskind** [comment]
I opened PR #26 to solve this and issue #12  .
Thank you to @mpoeter for contributing!
